lightly_studio/tests/resolvers/test_twodim_embedding_resolver.py

```diff
@@ -1,13 +1,155 @@
+    assert sample_ids_first_call == sample_ids_second_call
+
+    # Third call after adding a sample without embeddings - should still use cache.
+    helpers_resolvers.create_sample(
+        session=test_db,
+        dataset_id=dataset.dataset_id,
+        file_path_abs="/sample_2.jpg",
+    )
+    x_third, y_third, sample_ids_third_call = twodim_embedding_resolver.get_twodim_embeddings(
+        session=test_db,
+        dataset_id=dataset.dataset_id,
+        embedding_model_id=embedding_model.embedding_model_id,
+    )
+    calculate_spy.assert_called_once()
+    np.testing.assert_allclose(x_first, x_third)
+    np.testing.assert_allclose(y_first, y_third)
+    assert sample_ids_first_call == sample_ids_third_call
+
+
+def test__get_twodim_embeddings__recomputes_when_samples_change(
+    test_db: Session,
+    mocker: MockerFixture,
+) -> None:
+    # Create dataset, embedding model, samples, and embeddings.
+    dataset = helpers_resolvers.create_dataset(session=test_db, dataset_name="cache_miss_dataset")
+    embedding_model = helpers_resolvers.create_embedding_model(
+        session=test_db,
+        dataset_id=dataset.dataset_id,
+        embedding_dimension=3,
+    )
+
+    first_sample = helpers_resolvers.create_sample(
+        session=test_db,
+        dataset_id=dataset.dataset_id,
+        file_path_abs="/sample_0.jpg",
+    )
+    helpers_resolvers.create_sample_embedding(
+        session=test_db,
+        sample_id=first_sample.sample_id,
+        embedding_model_id=embedding_model.embedding_model_id,
+        embedding=[0.1, 0.1, 0.1],
+    )
+
+    calculate_spy = mocker.spy(twodim_embedding_resolver, "_calculate_2d_embeddings")
+
+    # First call - should call _calculate_2d_embeddings.
+    x_first, y_first, sample_ids_first = twodim_embedding_resolver.get_twodim_embeddings(
+        session=test_db,
+        dataset_id=dataset.dataset_id,
+        embedding_model_id=embedding_model.embedding_model_id,
+    )
+
+    assert calculate_spy.call_count == 1
+    assert x_first.shape == (1,)
+    assert y_first.shape == (1,)
+    assert len(sample_ids_first) == 1
+
+    # Add another sample and embedding.
+    second_sample = helpers_resolvers.create_sample(
+        session=test_db,
+        dataset_id=dataset.dataset_id,
+        file_path_abs="/sample_1.jpg",
+    )
+    helpers_resolvers.create_sample_embedding(
+        session=test_db,
+        sample_id=second_sample.sample_id,
+        embedding_model_id=embedding_model.embedding_model_id,
+        embedding=[0.2, 0.2, 0.2],
+    )
+
+    # Second call - should recompute since samples changed.
+    x_second, y_second, sample_ids_second = twodim_embedding_resolver.get_twodim_embeddings(
+        session=test_db,
+        dataset_id=dataset.dataset_id,
+        embedding_model_id=embedding_model.embedding_model_id,
+    )
+
+    assert calculate_spy.call_count == 2
+    assert x_second.shape == (2,)
+    assert y_second.shape == (2,)
+    assert len(sample_ids_second) == 2
```

Comment:
I noticed we do not have tests for the two branches that return `empty, empty, []`. Would it be possible to add a test?